ethcore/src/engines/clique/block_state.rs

```diff
@@ -0,0 +1,350 @@
+// Copyright 2015-2019 Parity Technologies (UK) Ltd.
+// This file is part of Parity Ethereum.
+
+// Parity Ethereum is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+
+// Parity Ethereum is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with Parity Ethereum.  If not, see <http://www.gnu.org/licenses/>.
+
+use std::collections::{HashMap, BTreeSet, VecDeque};
+use std::fmt;
+use std::time::{Duration, SystemTime, UNIX_EPOCH};
+
+use engines::EngineError;
+use engines::clique::util::{extract_signers, recover_creator};
+use engines::clique::{VoteType, DIFF_INTURN, DIFF_NOTURN, NULL_AUTHOR, SIGNING_DELAY_NOTURN_MS};
+use error::{Error, BlockError};
+use ethereum_types::{Address, H64};
+use rand::Rng;
+use types::BlockNumber;
+use types::header::Header;
+use unexpected::Mismatch;
+
+/// Type that keeps track of the state for a given vote
+// Votes that go against the proposal aren't counted since it's equivalent to not voting
+#[derive(Copy, Clone, Debug, PartialEq, PartialOrd)]
+pub struct VoteState {
+	kind: VoteType,
+	votes: u64,
+}
+
+/// Type that represent a vote
+#[derive(Copy, Clone, Debug, PartialEq, PartialOrd)]
+pub struct Vote {
+	block_number: BlockNumber,
+	beneficiary: Address,
+	kind: VoteType,
+	signer: Address,
+	reverted: bool,
+}
+
+/// Type that represent a pending vote
+#[derive(Copy, Clone, Debug, Eq, Hash, PartialEq, PartialOrd)]
+pub struct PendingVote {
+	signer: Address,
+	beneficiary: Address,
+}
+
+/// Clique state for each block.
+#[cfg(not(test))]
+#[derive(Clone, Debug, Default)]
+pub struct CliqueBlockState {
+	/// Current votes for a beneficiary
+	votes: HashMap<PendingVote, VoteState>,
+	/// A list of all votes for the given epoch
+	votes_history: Vec<Vote>,
+	/// a list of all valid signer, sorted by ascending order.
+	signers: BTreeSet<Address>,
+	/// a deque of recent signer, new entry should be pushed front, apply() modifies this.
+	recent_signers: VecDeque<Address>,
+	/// inturn signing should wait until this time
+	pub next_timestamp_inturn: Option<SystemTime>,
+	/// noturn signing should wait until this time
+	pub next_timestamp_noturn: Option<SystemTime>,
+}
+
+#[cfg(test)]
+#[derive(Clone, Debug, Default)]
+pub struct CliqueBlockState {
```

Comment:
`cfg(test)` and `cfg(not(test))` is exactly the same?

Reply:
Not exactly, all fields are public in the test version vs only `next_timestamp_noturn` and `next_timestamp_inturn` are pub